src/Authoring/WinRT.SourceGenerator/DiagnosticUtils.cs

```diff
@@ -0,0 +1,619 @@
+                case "IndexerDeclarationSyntax":
+                    return "indexer";
+                case "MethodDeclarationSyntax":
+                    return "method";
+                case "OperatorDeclarationSyntax":
+                    return "operator";
+                case "PropertyDeclarationSyntax":
+                    return "property";
+                default:
+                    return "unknown syntax type: " + syntaxType;
+            }
+        }
+
+        /// <summary>
+        /// returns true iff there is a field of the given type in the given struct 
+        /// e.g., if T is PropertyDeclarationSyntax, then if the struct has a property, we report a diagnostic and return true */
+        /// </summary>
+        /// <typeparam name="T">T can vary over MethodDeclartion, EventDeclaration, etc... </typeparam>
+        /// <param name="context"></param>
+        /// <param name="structDeclaration"></param>
+        /// <returns></returns>
+        public bool StructHasFieldOfType<T>(ref GeneratorExecutionContext context, StructDeclarationSyntax structDeclaration)
+        {
+            if (structDeclaration.DescendantNodes().OfType<T>().Any())
+            {
+                context.ReportDiagnostic(Diagnostic.Create(DiagnosticRules.StructHasInvalidFieldRule2, 
+                    structDeclaration.GetLocation(), 
+                    structDeclaration.Identifier, 
+                    SimplifySyntaxTypeString(typeof(T).Name)));
+                return true;
+            }
+            return false;
+        }
+
+        /// <summary>
+        /// returns true if there is a field declared private, 
+        /// or (inclusive) declared with a type that is a class or one of object, byte or dynamic
+        /// </summary> 
+        /// <returns>True if the struct has a field of an type that is not supported in Windows Runtime</returns>
+        public bool StructHasFieldOfInvalidType(ref GeneratorExecutionContext context, 
+            FieldDeclarationSyntax field, 
+            StructDeclarationSyntax structDeclaration, 
+            List<string> classNames)
+        {
+            bool found = false; 
+
+            /* No private fields allowed in Windows Runtime components */
+            if (!ModifiersContains(field.Modifiers, "public"))
+            { 
+                context.ReportDiagnostic(Diagnostic.Create(DiagnosticRules.StructHasPrivateFieldRule, field.GetLocation(), structDeclaration.Identifier));
+                found |= true;
+            }
+
+            if (ModifiersContains(field.Modifiers, "const"))
+            {
+                context.ReportDiagnostic(Diagnostic.Create(DiagnosticRules.StructHasConstFieldRule, field.GetLocation(), structDeclaration.Identifier));
+                found |= true;
+            }
+
+            foreach (var variable in field.DescendantNodes().OfType<VariableDeclarationSyntax>())
+            {
+                var typeStr = variable.Type.ToString();

```

Comment:
Can this be a TypeKind check whether we check like variable.Type.TypeKind == class or something.  This is assuming no classes are allowed.

Reply:
Unfortunately, at this point in the analysis the type is "IdentifierSyntax", because it is not a known type it is considered an identifier for a variable.

Reply:
Related to the `INamedTypeSymbol` comment, we might be able to do something similar here

Reply:
Discussed offline -- we are catching qualified and unqualified here now